File: java/src/main/java/com/logicalclocks/featurestore/metadata/Join.java
package com.logicalclocks.featurestore.metadata;

import com.logicalclocks.featurestore.Feature;
import com.logicalclocks.featurestore.JoinType;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

public class Join {

  @Getter @Setter
  private Query query;

  @Getter @Setter
  private List<Feature> on;
  @Getter @Setter
  private List<Feature> leftOn;
  @Getter @Setter
  private List<Feature> rightOn;

  @Getter @Setter
  private JoinType joinType;

  public Join(Query query, JoinType joinType) {
    this.query = query;
    this.joinType = joinType;
  }

  public Join(Query query, List<Feature> on, JoinType joinType) {
    this.query = query;
    this.on = on;
    this.joinType = joinType;
  }

  public Join(Query query, List<Feature> leftOn, List<Feature> rightOn, JoinType joinType) {
    this.query = query;
    this.leftOn = leftOn;
    this.rightOn = rightOn;
    this.joinType = joinType;
  }
}

File: python/hopsworks/core/join.py
from hopsworks import util


class Join:
    INNER = "INNER"
    LEFT = "LEFT"
    RIGHT = "RIGHT"
    FULL = "FULL"
    CROSS = "CROSS"
    LEFT_SEMI_JOIN = "LEFT_SEMI_JOIN"
    COMMA = "COMMA"

    def __init__(self, query, on, left_on, right_on, join_type):

        self._query = query
        self._on = util.parse_features(on)
        self._left_on = util.parse_features(left_on)
        self._right_on = util.parse_features(right_on)
        self._join_type = join_type or self.INNER
